File: plugin/src/main/java/co/uk/magmo/puretickets/storage/MySQL/MySQLManager.java
package co.uk.magmo.puretickets.storage.MySQL;

import co.uk.magmo.puretickets.storage.FunctionInterfaces.MessageFunctions;
import co.uk.magmo.puretickets.storage.FunctionInterfaces.MiscellaneousFunctions;
import co.uk.magmo.puretickets.storage.FunctionInterfaces.TicketFunctions;
import co.uk.magmo.puretickets.storage.SQLManager;
import co.uk.magmo.puretickets.storage.TimeAmount;
import co.uk.magmo.puretickets.ticket.Message;
import co.uk.magmo.puretickets.ticket.Ticket;
import co.uk.magmo.puretickets.ticket.TicketStatus;
import org.bukkit.Location;
import org.bukkit.plugin.Plugin;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.UUID;

public class MySQLManager extends SQLManager {
    public MySQLManager() {
        super(new MiscellaneousImpl(), );
    }

    private static class MiscellaneousImpl implements MiscellaneousFunctions {
        @Override
        public void setup(Plugin plugin) {

        }

        @Override
        public HashMap<UUID, Integer> highscores(TimeAmount amount) {
            return null;
        }
    }

    private static class TicketImpl implements TicketFunctions {
        @Override
        public co.uk.magmo.puretickets.ticket.Ticket select(Integer id) {
            return null;
        }

        @Override
        public List<co.uk.magmo.puretickets.ticket.Ticket> selectAll(TicketStatus status) {
            return null;
        }

        @Override
        public List<co.uk.magmo.puretickets.ticket.Ticket> selectAll(UUID uuid, TicketStatus status) {
            return null;
        }

        @Override
        public List<Integer> selectIds(UUID uuid, TicketStatus status) {
            return null;
        }

        @Override
        public Integer selectHighestId(UUID uuid, TicketStatus... statuses) {
            return null;
        }

        @Override
        public List<String> selectNames(TicketStatus status) {
            return null;
        }

        @Override
        public HashMap<TicketStatus, Integer> selectTicketStats(UUID uuid) {
            return null;
        }

        @Override
        public Boolean exists(Integer id) {
            return null;
        }

        @Override
        public Integer count(TicketStatus status) {
            return null;
        }

        @Override
        public Integer insert(UUID uuid, TicketStatus status, UUID picker, Location location) {
            return null;
        }

        @Override
        public void update(co.uk.magmo.puretickets.ticket.Ticket ticket) {

        }
    }

    private static class MessageImpl implements MessageFunctions {
        @Override
        public ArrayList<co.uk.magmo.puretickets.ticket.Message> selectAll(Integer id) {
            return null;
        }

        @Override
        public void insert(co.uk.magmo.puretickets.ticket.Ticket ticket, co.uk.magmo.puretickets.ticket.Message message) {

        }
    }
}

File: plugin/src/main/java/co/uk/magmo/puretickets/storage/MySQLManager.kt
package co.uk.magmo.puretickets.storage

import co.aikar.idb.*
import co.uk.magmo.puretickets.configuration.Config
import co.uk.magmo.puretickets.utils.asName
import org.bukkit.plugin.Plugin
import java.util.logging.Level
import kotlin.collections.HashMap

class MySQLManager : SQLManager {
    override fun setup(plugin: Plugin) {
        val options = DatabaseOptions.builder().mysql(Config.storageUser, Config.storagePassword, Config.storageName, Config.storageHost).build()
        val pooledOptions = PooledDatabaseOptions.builder().options(options).build()

        pooledOptions.dataSourceProperties = hashMapOf<String, Any>(Pair("useSSL", Config.storageSSL))

        val database = HikariPooledDatabase(pooledOptions)

        DB.setGlobalDatabase(database)
        DB.executeUpdate("CREATE TABLE IF NOT EXISTS puretickets_ticket(id INTEGER, uuid TEXT, status TEXT, picker TEXT)")
        DB.executeUpdate("CREATE TABLE IF NOT EXISTS puretickets_message(ticket INTEGER, reason TEXT, data TEXT, sender TEXT, date TEXT)")
        DB.executeUpdate("CREATE TABLE IF NOT EXISTS puretickets_notification(uuid TEXT, message TEXT, replacements TEXT)")
        DB.executeUpdate("CREATE TABLE IF NOT EXISTS puretickets_settings(uuid TEXT, announcements TEXT)")
        DB.executeUpdate("CREATE TABLE IF NOT EXISTS puretickets_sql(version INTEGER)")

        var version = DB.getFirstColumn<Int>("SELECT version FROM puretickets_sql")

        if (version == null) {
            DB.executeInsert("INSERT INTO puretickets_sql(version) VALUES(?)", 0)
            version = 0
        }

        if (version == 0) {
            plugin.logger.log(Level.INFO, "Updated PureTickets database to have location column")
            DB.executeUpdate("ALTER TABLE puretickets_ticket ADD location TEXT")
            version++
        }

        if (version <= 1) {
            plugin.logger.log(Level.INFO, "Updated PureTickets database to remove tickets with empty locations and remove all pending notifications")
            DB.executeUpdate("DELETE from puretickets_ticket WHERE location IS NULL OR trim(location) = ?", "")
            DB.executeUpdate("DELETE from puretickets_notification")
            version++
        }

        DB.executeUpdate("UPDATE puretickets_sql SET version = ?", version)
    }

    override fun DbRow.getPureLong(column: String): Long {
        return getString(column).toLong()
    }
}